include/cuco/detail/bloom_filter/bloom_filter_policy_impl.cuh

```diff
@@ -0,0 +1,120 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuco/hash_functions.cuh>
+
+// TODO switch to <cuda/std/algorithm> once available
+#include <cuda/std/__algorithm/max.h>
+#include <cuda/std/__algorithm/min.h>
+#include <cuda/std/array>
+#include <cuda/std/bit>
+#include <cuda/std/limits>
+#include <cuda/std/tuple>
+#include <cuda/std/type_traits>
+
+#include <cstdint>
+
+namespace cuco::detail {
+
+template <class Hash, class Block>
+class bloom_filter_policy_impl {
+ public:
+  using hasher             = Hash;
+  using word_type          = typename Block::value_type;
+  using hash_argument_type = typename hasher::argument_type;
+  using hash_result_type   = decltype(std::declval<hasher>()(std::declval<hash_argument_type>()));
+
+  static constexpr std::uint32_t words_per_block = cuda::std::tuple_size_v<Block>;
+
+ private:
+  static constexpr std::uint32_t word_bits =
+    cuda::std::numeric_limits<typename Block::value_type>::digits;
+  static constexpr std::uint32_t bit_index_width = cuda::std::bit_width(word_bits - 1);
+
+ public:
+  static_assert(cuda::std::has_single_bit(words_per_block) and words_per_block <= 32,
+                "Number of words per block must be a power-of-two and less than or equal to 32");
+
+  __host__ __device__ constexpr bloom_filter_policy_impl(uint32_t pattern_bits, Hash hash)
+    : hash_{hash}
+  {
+    constexpr uint32_t hash_bits = cuda::std::numeric_limits<hash_result_type>::digits;
+
+    // This ensures each word in the block has at least one bit set; otherwise we would never use
+    // some of the words
+    constexpr uint32_t min_pattern_bits = words_per_block;
+
+    // The maximum number of bits to be set for a key is capped by the number of bits in the hash
+    // value or the number of bits in the filter block
+    constexpr uint32_t max_pattern_bits =
+      cuda::std::min(hash_bits / bit_index_width, word_bits * words_per_block);
+
+    // TODO do we want this to fail?
+    // constexpr uint32_t max_pattern_bits_from_hash = hash_bits / bit_index_width;
+    // static_assert(min_pattern_bits <= max_pattern_bits_from_hash, "Hash value too narrow for the
+    // specified filter block type");
+
+    // TODO update this part once cuda::std::clamp is available
+    // Here we adjust the value dynamically; alternatively we could throw an error if the value is
+    // outside the expected range
+    pattern_bits_ =
+      cuda::std::min(cuda::std::max(min_pattern_bits, pattern_bits), max_pattern_bits);
+
+    min_bits_per_word_ = pattern_bits_ / words_per_block;
+    remainder_bits_    = pattern_bits_ % words_per_block;
+  }
+
+  __device__ constexpr hash_result_type hash(hash_argument_type const& key) const
+  {
+    return hash_(key);
+  }
+
+  template <class Extent>
+  __device__ constexpr auto block_index(hash_result_type hash, Extent num_blocks) const
+  {
+    return hash % num_blocks;
```

Comment:
`((uint64_t)hash * (uint64_t)num_blocks) >> 32` is usually much faster and is the way this is done in Parquet, IIRC.

Reply:
Wouldn't that restrict the number of blocks to 2^32, i.e., if `words_per_block == 1` then the filter can't be larger than 4GB (or at least we wouldn't use any blocks past that boundary).

We have two facilities in cuco to speed up the remainder computation:

- We can specify a compile-time extent which allows the compiler to apply a suitable optimization for the remainder computation. This should already work with the current implementation.
- If the extent is a runtime value, we can wrap it in `cuco::utility::fast_int` ([link](https://github.com/NVIDIA/cuCollections/blob/5602381ff980d3d665760e235e23b9daa368a65b/include/cuco/utility/fast_int.cuh#L35)) which provides optimized division and modulo operators. However, I need to figure out how to dispatch this code path when constructing the filter.

Reply:
Yes, it would. So for larger filters, one can use larger int types, including simulating them (which is still fast even with schoolbook long multiplication).